### PR TITLE
Use libc for `libdl` bindings on Linux

### PR DESCRIPTION
**Pull request by GabrielMajeri**
Since we don't need a lot of stuff from `link.h`, I think it's better to drop bindgen and manually define the things we need from the header.

Note: a similar thing could be done for MacOS, so we can drop bindgen there too. I don't have a Mac device to develop on however.

### CODE REVIEW

**Comment by philipc** [comment]
@GabrielMajeri What's your motivation for this? To make it compile faster?

@fitzgen I can give this a detailed review if you don't have time, but I'd like your approval that this is the right way to go first.

**Comment by GabrielMajeri** [comment]
@philipc To make it compile faster and reduce dependencies on `libclang` / bindgen. Should also fix #19.

The CI failed because it didn't manage to install Clang on MacOS. Considering we need only a small part of the headers, it seems better to just manually write a binding to what we need.

I also though of using `libc`'s bindings, but we'd need to pull in `libc` as a dependency and their structs don't yet have the `Debug` trait.

**Comment by fitzgen** [comment]
I empathize with the desire to avoid libclang and bindgen dependencies.

However, I would strongly prefer using `libc`'s bindings over bespoke hand-written equivalents, because `libc` has thorough testing with `ctest` that compares the sizes and layouts of the Rust structs with their C counterparts to ensure that they are correct.

As far as debug impls go, we can write our own by hand instead of using derive. It is a bit of boilerplate, but worth the trade off IMO.

**Comment by GabrielMajeri** [comment]
@fitzgen Done, now we use libc and implement debug manually. I don't know if my debug impl is 100% the same as before, but it displays all tthe information.

File: src/linux/mod.rs [hunk 1]
```diff
@@ -88,8 +87,18 @@ impl<'a> Iterator for SegmentIter<'a> {
     }
 }
 
+impl<'a> fmt::Debug for SegmentIter<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let ref phdr = self.inner.as_slice()[0];
+
+        write!(f, "SegmentIter {{ phdr: ")?;
+        fmt_debug_phdr(phdr, f)?;
+        write!(f, " }}")
```
Comment (fitzgen) [inline src/linux/mod.rs]: Nitpick: This should be 

```rust
f.debug_struct().field("phdr", &DebugPhdr(phd)).finish()
```

File: src/linux/mod.rs [hunk 2]
```diff
@@ -185,6 +194,34 @@ impl<'a> SharedLibraryTrait for SharedLibrary<'a> {
...
+}
+
+// Helper function for implementing Debug for structures which contain Phdr.
+fn fmt_debug_phdr(phdr: &Phdr, f: &mut fmt::Formatter) -> fmt::Result {
+    // The layout is different for 32-bit vs 64-bit, but since the fields are the same,
+    // it shouldn't matter much.
+    write!(f, "Phdr {{ p_type: {:#X}, p_flags: {:?}, ", phdr.p_type, phdr.p_flags)?;
+    write!(f, "p_offset: {:#X}, p_vaddr: {:#X}, ", phdr.p_offset, phdr.p_vaddr)?;
+    write!(f, "p_paddr: {:#X}, p_filesz: {:?}, ", phdr.p_paddr, phdr.p_filesz)?;
+    write!(f, "p_memsz: {:?}, p_align: {:#X} }}", phdr.p_memsz, phdr.p_align)
```
Comment (fitzgen) [inline src/linux/mod.rs]: And then this whole thing should be

```rust
DebugPhdr<'a>(&'a Phdr);

impl<'a> fmt::Debug for DebugPhdr<'a> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.debug_struct("Phdr")
            .field("p_type", &self.0.p_type)
            // etc...
            .finish()
    }
}
```

File: src/linux/mod.rs (hunk 1 above)
Comment (GabrielMajeri) [reply src/linux/mod.rs]: Cool, I guess I learn something new everyday.

**Pull request merged**